### PR TITLE
Refactor RenameDriver to use derived record type

### LINKED ISSUE #590: FieldRenamer doesn't rename lazy records
The following unit test fails:

        private class LazyRecursiveRecordType : RecordType
        {
            public override IEnumerable<string> FieldNames => GetFieldNames();

            public bool EnumerableIterated = false;

            public LazyRecursiveRecordType()
                : base()
            {
            }

            public override bool TryGetFieldType(string name, out FormulaType type)
            {
                switch (name)
                {
                    case "SomeString":
                        type = FormulaType.String;
                        return true;
                    case "Loop":
                        type = new LazyRecursiveRecordType();
                        return true;
                    default:
                        type = FormulaType.Blank;
                        return false;
                }
            }

            private IEnumerable<string> GetFieldNames()
            {
                EnumerableIterated = true;

                yield return "SomeString";
                yield return "Loop";
            }

            public override bool Equals(object other)
            {
                return other is LazyRecursiveRecordType; // All the same 
            }

            public override int GetHashCode()
            {
                return 1;
            }
        }

        [Fact]
        public void Rename()
        {
            var engine = new Engine(new PowerFxConfig(CultureInfo.InvariantCulture));

            var dpath = DPath.Root;
            foreach (var segment in "Loop.SomeString".Split('.'))
            {
                dpath = dpath.Append(new DName(segment));
            }

            var renamer = engine.CreateFieldRenamer(new LazyRecursiveRecordType(), dpath, new DName("Var"));

            Assert.Equal("Loop.Var = \"1\"", renamer.ApplyRename("Loop.SomeString = \"1\""));
        }

### PR DESCRIPTION
**Pull request by lesaltzm**
fixes #590 

In adding support for #590, this also refactors RenameDriver to be significantly lazier. It no longer needs to traverse the entire parameters input, just the portion actually accessed in the expression. This is done by leveraging the new lazy type support to wrap the parameter record and lazily replace the display name providers on demand.

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by MikeStall — APPROVED** [review]
:shipit:

File: src/libraries/Microsoft.PowerFx.Core/Public/DisplayNames/RenameDriver.cs [hunk 1]
```diff
@@ -61,23 +66,97 @@ private static FormulaType RenameFormulaTypeHelper(AggregateType nestedType, Que
...
-            // Use DType internals to swap one field type for the updated one and disable all other display names
-            var dropped = DType.DisableDisplayNameProviders(nestedType._type.Drop(ref fError, DPath.Root, field));
-            Contracts.Assert(!fError);
+            // Wrap the nestedType, swapping one field for the updated one and disabling all other display names
+            return GetWrappedAggregateType(nestedType, DisabledDisplayNameProvider.Instance, field, innerUpdatedType);
+        }
+
+        private AggregateType GetWrappedAggregateType(AggregateType rootType, DisplayNameProvider displayNameProvider, string replacedFieldName = null, FormulaType replacedFieldType = null)
+        {
+            if (!rootType._type.IsLazyType)
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Public/DisplayNames/RenameDriver.cs:86]: >_type [](http://example.com/codeflow?start=26&length=5)

Is "_" used even for internal type in naming? I thought that this was only used for private members.

Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Core/Public/DisplayNames/RenameDriver.cs:86]: No, this should be changed at some point

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by MikeStall — APPROVED** [review]

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Pull request merged**